doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,879 @@
+>>> versin(x)
+versin(x)
+>>> isinstance(versin(x), versin)
+True
+```
+
+```{note}
+
+All the methods described below are optional. They can be included if you want
+to define the given behavior, but if they are omitted, SymPy will default to
+leaving things unevaluated. For example, if you do not define
+[differentiation](custom-functions-differentiation), {func}`~.diff` will just
+return an unevaluated {class}`~.Derivative`.
+
+```
+
+(custom-functions-eval)=
+### Defining Automatic Evaluation with `eval`
+
+```{sidebar} Reminder
+Remember that `eval` should be defined with the `@classmethod` decorator.
+```
+
+The first and most common thing we might want to define on our custom function
+is automatic evaluation, that is, the cases where it will return an actual
+value instead of just remaining unevaluated as-is.
+
+This is done by defining the class method `eval`. `eval` should take the
+arguments of the function and return either a value or `None`. If it returns
+`None`, the function will remain unevaluated in that case.
+
+For our function `versin`, we might recall that $\cos(n\pi) = (-1)^n$ for
+integer $n$, so $\operatorname{versin}(n\pi) = 1 - (-1)^n.$ We can make
+`versin` automatically evaluate to this value when passed an integer multiple
+of `pi`:
+
+(custom-functions-versin-eval-example)=
+```py
+>>> from sympy import pi, Integer
+>>> class versin(Function):
+...    @classmethod
+...    def eval(cls, x):
+...        # If x is an integer multiple of pi, x/pi will cancel and be an Integer
+...        n = x/pi
+...        if isinstance(n, Integer):
+...            return 1 - (-1)**n
+```
+
+```py
+>>> versin(pi)
+2
+>>> versin(2*pi)
+0
+>>> versin(x*pi)
+versin(pi*x)
+```
+
+Here we make use of the fact that if a Python function does not explicitly
+return a value, it automatically returns `None`, so in the cases where the `if
+isinstance(n, Integer)` statement is not triggered, `eval` returns `None` and
+`versin` remains unevaluated.
+
+```{note}
+
+`Function` subclasses should not redefine `__new__` or `__init__`. If you want
+to implement behavior that isn't possible with `eval`, it might make more
+sense to subclass {class}`~.Expr` rather than `Function`.
+
+```
+
+`eval` can take any number of arguments, including an arbitrary number with
+`*args` and optional keyword arguments. The `.args` of the function will
+always be the arguments that were passed in by the user. For example
+
+```py
+>>> class f(Function):
+...     @classmethod
+...     def eval(cls, x, y=1, *args):
+...         return None
+```
+
+```py
+>>> f(1).args
+(1,)
+>>> f(1, 2).args
+(1, 2)
+>>> f(1, 2, 3).args
+(1, 2, 3)
+```
+
+Finally, note that automatic evaluation on floating-point inputs happens
+automatically once `evalf` is defined. See the section on [defining numerical
+evaluation with `evalf`](custom-functions-evalf) below.
+
+(custom-functions-eval-best-practices)=
+#### Best Practices for `eval`
+
+- **Don't just return an expression.**
```

Comment:
Per [SymPy tone preferences](https://docs.sympy.org/dev/guides/contributing/documentation-style-guide.html#tone-preferences), suggest phrasing this positively. For example, "Return _description_ instead of an expression".